Extract method view_to_association

moolitayer opened the pull request.

moolitayer commented:
cc @jrafanie

abonas commented on app/helpers/application_helper.rb:
```diff
@@ -260,6 +244,18 @@ def view_to_url(view, parent = nil)
 
+  def class_to_db(cname, parent)
+    return unless cname
+    case
```
afaik this can be shortened by doing `case cname` and removing repeating `cname` after each 'when' keyword

jrafanie replied:
Yes!  That's right.  That would then change line 253 and 254 below to `when "SystemService"` where we'd have to move subcondition into that single "when" of the case statement.

chessbyte replied:
@jrafanie that is how the original code had it, so this refactoring may have changed behavior without realizing it

jrafanie replied:
I agree, that's why I like @abonas' suggestion.  It's more of an extract method.

chessbyte commented on app/helpers/application_helper.rb:
```diff
@@ -260,6 +244,18 @@ def view_to_url(view, parent = nil)
+    when cname == "OrchestrationStackOutput" then "outputs"
+    when cname == "OrchestrationStackParameter" then "parameters"
+    when cname == "OrchestrationStackResource" then "resources"
+    when cname == "SystemService" && parent.class.base_class.to_s.downcase == "host" then "host_services"
```
why not `parent.kind_of?(Host)` instead of `parent.class.base_class.to_s.downcase == "host"`

chessbyte commented on app/helpers/application_helper.rb:
```diff
@@ -260,6 +244,18 @@ def view_to_url(view, parent = nil)
+    when cname == "OrchestrationStackParameter" then "parameters"
+    when cname == "OrchestrationStackResource" then "resources"
+    when cname == "SystemService" && parent.class.base_class.to_s.downcase == "host" then "host_services"
+    when cname == "SystemService" && parent.class.base_class.to_s.downcase == "vm"   then @lastaction
```
why not `parent.kind_of?(Vm)` instead of `parent.class.base_class.to_s.downcase == "vm"`

jrafanie commented on app/helpers/application_helper.rb:
```diff
@@ -199,24 +199,8 @@ def url_for_db(db, action = "show", item = nil) # Default action is show
-      when "vm"   then association = @lastaction
-      end
-    end
+    association = class_to_db(view.db, parent)
```
I don't know about the name `class_to_db`...   I don't know what db means.  Is it really an association? If so, maybe view_to_association or something similar would be better.

moolitayer commented:
- addressed comments
- added another commit renaming another db_to_x method: db_to_controller  to view_to_controller (consistently with view_to_association)

Personally I think taking comments into account the best impl for view_to_association will be:
```
def view_to_association(cname, parent)
  return unless cname
  case
  when cname == "OrchestrationStackOutput" then "outputs"
  when cname == "OrchestrationStackParameter" then "parameters"
  when cname == "OrchestrationStackResource" then "resources"
  when cname == "SystemService" && parent.kind_of?(Vm) then @lastaction
  when cname == "SystemService" && parent.kind_of?(Host) then "host_services"
  else cname.tableize
  end
end
```
- I prefer a complex condition to nesting
- I think the test covers the use cases but I can't be sure.

@jrafanie @abonas @chessbyte please review

moolitayer commented:
@jrafanie @abonas @chessbyte [what say you?](https://github.com/ManageIQ/manageiq/pull/7259#issuecomment-196800633) which way is better?

jrafanie commented:
I'd rather have a view type object that you ask for it's association or better understand what this is used for...a url builder, so maybe association_url? but for the purpose of this pr, I'm good with the nested conditional over a repetitive case statement  with a complex conditional.

chessbyte commented on app/helpers/application_helper.rb:
```diff
@@ -199,26 +199,10 @@ def url_for_db(db, action = "show", item = nil) # Default action is show
-      when "vm"   then association = @lastaction
-      end
-    end
+    association = view_to_association(view.db, parent)
```
Why set `association` twice?  Old code was setting an initial value and overriding it under certain conditions.  New code is always setting it to whatever the `view_to_association` method returns.

moolitayer replied:
oh my :open_mouth:  

Also I had hoped this patch would mean I won't have to touch this code [here](https://github.com/ManageIQ/manageiq/pull/6839/files#diff-127a850b7d7b20173ad3fba88f8a25e3R205) but I can't think of how to do that now.

Anyway I don't know if this patch is still worth the risk, it but I gave it another go.
Please review @chessbyte @jrafanie @abonas

jrafanie commented:
@moolitayer can you make sure you renamed db_to_controller to view_to_controller everywhere?

My git grep below shows some in application_helper_spec:
```
app/helpers/application_helper.rb:198:      controller, action = db_to_controller(db, action)
app/helpers/application_helper.rb:226:      controller, action = db_to_controller(view.db)
app/helpers/application_helper.rb:269:  def db_to_controller(db, action = "show")
spec/helpers/application_helper_spec.rb:247:        c, a = helper.db_to_controller(db, @action)
spec/helpers/application_helper_spec.rb:268:        c, a = helper.db_to_controller(db, @action)
spec/helpers/application_helper_spec.rb:279:  describe "#db_to_controller" do
spec/helpers/application_helper_spec.rb:280:    subject { helper.db_to_controller(@db) }
```

dclarizio commented:
@moolitayer need to address a bunch tests failing similar to:
```
 5) ApplicationHelper#db_to_controller when with EmsInfra and @explorer

     Failure/Error: subject { helper.db_to_controller(@db) }

     NoMethodError:

       undefined method `db_to_controller' for #<#<Class:0x0000001f628648>:0x000000382204d8>
```

moolitayer commented:
@jrafanie @dclarizio  thanks
renamed db_to_controller to view_to_controller everywhere

dclarizio commented:
@moolitayer So looking at all the places `view_to_controller` is being called, I'm wondering why the name change since we always seem to be passing it a `db`, not a `view`.  I'm not arguing for or against, just wonder what the logic is driving the change.

moolitayer commented:
> @moolitayer So looking at all the places view_to_controller is being called, I'm wondering why the name change since we always seem to be passing it a db, not a view. I'm not arguing for or against, just wonder what the logic is driving the change.

Probably me writing the patch late at night :eyes: 
I removed the second commit.

chessbyte commented on spec/helpers/application_helper_spec.rb:
```diff
@@ -1628,4 +1628,31 @@
     end
   end
+
+  describe '.view_to_table' do
```
should this be `'.view_to_association'`?  Is it a class or instance method?  If instance method, then `describe '#view_to_association'`.

chessbyte commented on app/helpers/application_helper.rb:
```diff
@@ -265,6 +247,20 @@ def view_to_url(view, parent = nil)
+  def view_to_association(view, parent)
+    case view.db
+    when "OrchestrationStackOutput", "OrchestrationStackParameter", "OrchestrationStackResource"
+                                       then view.db.split(/(?=[A-Z])/).last.downcase.pluralize
```
is this processing really clearer than the original?

jrafanie replied:
@chessbyte I agree, I'd rather we have separate when statements for each one.

These non-standard view -> association names should really stick out so we can really understand why they're different.  In the original code, we broke the convention here and there's no explanation why... I'm sure this and other conditionals would just disappear if we stuck to conventions.

chessbyte commented on app/helpers/application_helper.rb:
```diff
@@ -265,6 +247,20 @@ def view_to_url(view, parent = nil)
+      when Host then "host_services"
+      when Vm   then @lastaction
+      end
+    end || view.scoped_association
```
should `view.scoped_association` be the `else` of the `case` statement?

jrafanie replied:
Yes, I think that's what the `||` is supposed to do but it's a very uncommon syntax if it works this way.

chessbyte commented:
@moolitayer It seems that this PR wants to be 2 PRs.  
One for the method extraction (and, thereby, easier testability) and another for the method refactoring.

chessbyte commented:
Green on [Travis](https://travis-ci.org/ManageIQ/manageiq/builds/118073457).  Merging...
